public/scripts/models/image-upload.js

```diff
@@ -45,23 +39,75 @@ class ImageUpload {
     });
     this.file = file;
     this.maxFileSizeByte = maxFileSizeByte;
-    if (!this.supportedFileTypes.includes(file.type)) {
+  }
+
+  async read() {
+    this.readState = ReadState.Reading;
+    if (!(await this._verifyFileType())) {
       this.readState = ReadState.ErrUnsupportedFileType;
       return;
     }
-    if (this.maxFileSizeByte != null && this.file.size > this.maxFileSizeByte) {
+    if (!this._verifyFileSize()) {
       this.readState = ReadState.ErrExceedMaxFileSize;
       return;
     }
-    this.readState = ReadState.Reading;
-    this.loadDimensionPromise = this._loadDimension();
+    const loadDimensionResult = await this._loadDimension();
+    if (loadDimensionResult.type === "failed") {
+      this.readState = loadDimensionResult.reason;
+      return;
+    }
+    this.readState = ReadState.ReadSuccess;
+  }
+
+  // Cache file type from header such that no need to parse again
+  __fileTypeFromHeader = null;
+  async _getFileTypeFromHeader() {
+    if (this.__fileTypeFromHeader != null) {
+      return this.__fileTypeFromHeader;
+    }
+    const arrayBuffer = await this.file.arrayBuffer();
+    const headerFileTypeBuff = new Uint8Array(arrayBuffer).subarray(0, 4);
+    let fileTypeFromHeader = "";
+    for (let i = 0; i < headerFileTypeBuff.length; i++) {
+      fileTypeFromHeader += String.fromCharCode(
+        parseInt(headerFileTypeBuff[i].toString(16), 16),
+      );
+    }
+    this.__fileTypeFromHeader = fileTypeFromHeader;
+    return this.__fileTypeFromHeader;
+  }
+
+  async _isPsd() {
+    const mimeTypes = ["application/x-photoshop", "image/vnd.adobe.photoshop"];
+    const fileTypesFromHeader = [
+      "8BPS", // https://www.fileformat.info/format/psd/egff.htm
+    ];
+    return (
+      mimeTypes.includes(this.file.type) ||
+      // Windows workaround
+      // https://stackoverflow.com/questions/51724649/mime-type-of-file-returning-empty-in-javascript-on-some-machines
+      fileTypesFromHeader.includes(await this._getFileTypeFromHeader())
+    );
+  }
+
+  _isImg() {
+    const mimeTypes = ["image/jpeg", "image/png"];
+    return mimeTypes.includes(this.file.type);
+  }
+
+  async _verifyFileType() {
+    return (await this._isPsd()) || this._isImg();
```

Comment:
can reverse this as ` this._isImg() || (await this._isPsd());`? Seems currently jpg & png will fail to upload

https://github.com/oursky/mockuphone.com/assets/74223769/b029018b-e14a-4e82-b5cf-947bbc48e65f

Reply:
Not related to the order. :sosad: should be fixed